File: 0600_0799/leetcode-no783.cpp
//Definition for a binary tree node.
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    void dfs(vector<int>& res,TreeNode* root){
        res.push_back(root->val);
        if (root->left){
            dfs(res,root->left);
        }
        if (root->right){
            dfs(res,root->right);
        }
    }
    int minDiffInBST(TreeNode* root) {
        vector<int> ans;
        dfs(ans,root);
        sort(ans.begin(), ans.end());
        int min = INT_MAX;
        for (int i = 1;i<ans.size();i++){
            if ((ans[i] - ans[i - 1]) < min) min = ans[i] - ans[i - 1];
        }
        return min;
    }
};


int main(){
    Solution test;
    TreeNode node1(1, nullptr, nullptr);
    TreeNode node3(3, nullptr, nullptr);
    TreeNode node2(2, &node1, &node3);
    TreeNode node6(6, nullptr, nullptr);
    TreeNode node4(4, &node2, &node6);
    cout << test.minDiffInBST(&node4) << endl;
}
//
// Created by ASUS on 2023/6/26.
//

File: leetcode-cs/0700_0799/leetcode-no783.cs
// problem: https://leetcode.cn/problems/minimum-distance-between-bst-nodes/?envType=problem-list-v2&envId=3rBpGyas 

namespace leetcodeNo783;

using System;
using Utils;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    private int res = Int32.MaxValue;
    private int last = -1;

    public int MinDiffInBST(TreeNode root) {
        dfs(root);
        return res;
    }

    public void dfs(TreeNode root) {
        int cur = root.val;

        if (root.left != null) {
            dfs(root.left);
        }

        if (last >= 0) {
            res = int.Min(res, cur - last);
        }

        last = cur;
        if (root.right != null) {
            dfs(root.right);
        }
    }
}

public class Test {
    public static void Run() {
        var s = new Solution();
        Console.WriteLine(s.MinDiffInBST(TreeGenerator.StringToTreeNode("[4,2,6,1,3]")));
        var s2 = new Solution();
        Console.WriteLine(s2.MinDiffInBST(TreeGenerator.StringToTreeNode("[1,0,48,null,null,12,49]")));
    }
}

//
// Created By jing At 10/31/25 20:10:24
//